File: packages/bot/src/lib/evaluation-helpers.ts
import { isRankActionable, Rank } from '@vinto/shapes';
import { MCTSGameState, MCTSPlayerState } from './mcts-types';

/**
 * NEW: Evaluate toss-in potential
 *
 * This is the key to multi-step thinking!
 * We reward having matching cards because they enable cascading toss-ins
 */
export function evaluateTossInPotential(
  state: MCTSGameState,
  botPlayer: MCTSPlayerState
): number {
  // Build rank frequency map
  const rankCounts = new Map<Rank, number>();
  const rankValues = new Map<Rank, number>();

  for (let pos = 0; pos < botPlayer.cardCount; pos++) {
    const memory = botPlayer.knownCards.get(pos);
    if (memory && memory.confidence > 0.5 && memory.card) {
      const rank = memory.card.rank;
      rankCounts.set(rank, (rankCounts.get(rank) || 0) + 1);

      if (!rankValues.has(rank)) {
        rankValues.set(rank, memory.card.value);
      }
    }
  }

  let tossInValue = 0;

  // Calculate value of potential toss-ins
  rankCounts.forEach((count, rank) => {
    if (count >= 2) {
      // We have pairs/triples - HIGH value!
      const value = rankValues.get(rank) || 5;

      // Value formula: (count - 1) * value * multiplier
      // - Pairs worth 1x value
      // - Triples worth 2x value
      // - Quads worth 3x value
      const cascadeMultiplier = count - 1;
      tossInValue += cascadeMultiplier * value * 3; // 3x multiplier

      // EXTRA bonus for high-value pairs (removing 10s, Js, Qs, Ks)
      if (value >= 10) {
        tossInValue += cascadeMultiplier * 5;
      }

      // HUGE bonus for King pairs (enables King cascade + action)
      if (rank === 'K') {
        tossInValue += cascadeMultiplier * 10;
      }
    }
  });

  // Normalize to 0-1 (max potential toss-in value ≈ 60)
  return Math.min(1, tossInValue / 60);
}

/**
 * Evaluate relative position vs opponents
 */
export function evaluateRelativePosition(
  state: MCTSGameState,
  botPlayer: MCTSPlayerState
): number {
  const opponents = state.players.filter((p) => p.id !== botPlayer.id);

  const botScore = botPlayer.score;
  const opponentScores = opponents.map((p) => p.score);
  const avgOpponentScore =
    opponentScores.reduce((a, b) => a + b, 0) / opponentScores.length;
  const bestOpponentScore = Math.min(...opponentScores);

  // Score advantage
  const advantageVsAvg = avgOpponentScore - botScore;
  const advantageVsBest = bestOpponentScore - botScore;

  // Card count advantage
  const botCards = botPlayer.cardCount;
  const avgOpponentCards =
    opponents.reduce((sum, p) => sum + p.cardCount, 0) / opponents.length;
  const cardAdvantage = avgOpponentCards - botCards;

  // Normalize
  const scoreComponent = Math.min(1, Math.max(0, (advantageVsAvg + 10) / 30));
  const cardComponent = Math.min(1, Math.max(0, (cardAdvantage + 2) / 5));
  const competitiveComponent = Math.min(
    1,
    Math.max(0, (advantageVsBest + 5) / 25)
  );

  return (
    scoreComponent * 0.5 + cardComponent * 0.3 + competitiveComponent * 0.2
  );
}

/**
 * NEW: Evaluate action card value
 *
 * Having action cards (especially Kings) is valuable
 */
export function evaluateActionCardValue(
  state: MCTSGameState,
  botPlayer: MCTSPlayerState
): number {
  let actionValue = 0;

  const rankCounts = new Map<Rank, number>();

  for (let pos = 0; pos < botPlayer.cardCount; pos++) {
    const memory = botPlayer.knownCards.get(pos);
    if (memory && memory.confidence > 0.5 && memory.card) {
      const card = memory.card;
      const rank = card.rank;

      rankCounts.set(rank, (rankCounts.get(rank) || 0) + 1);

      // Action cards have strategic value
      if (isRankActionable(card.rank)) {
        if (rank === 'K') {
          actionValue += 15; // King is most powerful
        } else if (rank === 'Q' || rank === 'J') {
          actionValue += 10; // Queen/Jack very strong
        } else if (rank === '9' || rank === '10') {
          actionValue += 6; // Peek opponents
        } else if (rank === '7' || rank === '8') {
          actionValue += 4; // Peek own
        } else if (rank === 'A') {
          actionValue += 3; // Ace situational
        }
      }
    }
  }

  // SYNERGY BONUS: Having King + pairs enables powerful combos
  const hasKing = rankCounts.has('K');
  let pairCount = 0;
  rankCounts.forEach((count) => {
    if (count >= 2) pairCount++;
  });

  if (hasKing && pairCount > 0) {
    actionValue += pairCount * 8; // King + pairs = powerful synergy
  }

  // Normalize to 0-1 (max action value ≈ 50)
  return Math.min(1, actionValue / 50);
}

/**
 * Evaluate information advantage
 */
export function evaluateInformationAdvantage(
  state: MCTSGameState,
  botPlayer: MCTSPlayerState
): number {
  const ownKnowledge =
    botPlayer.cardCount > 0
      ? botPlayer.knownCards.size / botPlayer.cardCount
      : 0;

  // How much we know about opponents
  const opponents = state.players.filter((p) => p.id !== botPlayer.id);
  let totalOppCards = 0;
  let knownOppCards = 0;

  opponents.forEach((opp) => {
    totalOppCards += opp.cardCount;
    knownOppCards += opp.knownCards.size;
  });

  const oppKnowledge = totalOppCards > 0 ? knownOppCards / totalOppCards : 0;

  return ownKnowledge * 0.6 + oppKnowledge * 0.4;
}

/**
 * Evaluate threat level from opponents
 */
export function evaluateThreatLevel(
  state: MCTSGameState,
  botPlayer: MCTSPlayerState
): number {
  const opponents = state.players.filter((p) => p.id !== botPlayer.id);

  let threat = 0;

  opponents.forEach((opp) => {
    // Threat from fewer cards
    if (opp.cardCount < botPlayer.cardCount) {
      threat += (botPlayer.cardCount - opp.cardCount) * 0.1;
    }

    // Threat from lower score
    if (opp.score < botPlayer.score) {
      threat += (botPlayer.score - opp.score) / 40;
    }

    // Threat from unknown cards
    const unknownRatio =
      opp.cardCount > 0
        ? (opp.cardCount - opp.knownCards.size) / opp.cardCount
        : 0;
    threat += unknownRatio * 0.08;
  });

  // Invert (lower threat = higher score)
  return Math.max(0, 1 - threat);
}

File: kmp/shared/bot/src/commonMain/kotlin/game/vinto/bot/EvaluationHelpers.kt
package game.vinto.bot

import game.vinto.shapes.Rank
import game.vinto.shapes.isActionable

/**
 * The five things the search cares about when scoring a position, ported from
 * `packages/bot/src/lib/evaluation-helpers.ts`.
 *
 * Each returns 0-1 so they can be weighted against each other. The weights themselves live
 * in [evaluateState].
 */

/** Memories below this are hunches; the evaluator only counts what the bot is sure of. */

/**
 * Matching ranks in hand, which is where multi-step play comes from.
 *
 * A pair is not just two cards — it is a toss-in that removes both, and if they are Kings it
 * is a toss-in that also hands the bot an action. Rewarding pairs is what makes the search
 * look past the current turn without having to search deeper.
 */
fun evaluateTossInPotential(botPlayer: MctsPlayerState): Double {
    val rankCounts = mutableMapOf<Rank, Int>()
    val rankValues = mutableMapOf<Rank, Int>()

    for (position in 0 until botPlayer.cardCount) {
        val memory = botPlayer.knownCards[position] ?: continue
        if (memory.confidence <= TRUSTED_CONFIDENCE) continue

        val rank = memory.card.rank
        rankCounts[rank] = (rankCounts[rank] ?: 0) + 1
        rankValues.getOrPut(rank) { memory.card.value }
    }

    var tossInValue = 0.0
    for ((rank, count) in rankCounts) {
        if (count < 2) continue

        val value = rankValues[rank] ?: DEFAULT_UNKNOWN_VALUE
        // A pair sheds one extra card, a triple two, and so on.
        val cascade = count - 1

        tossInValue += cascade * value * CASCADE_MULTIPLIER
        if (value >= HIGH_VALUE_THRESHOLD) tossInValue += cascade * HIGH_VALUE_BONUS
        if (rank == Rank.KING) tossInValue += cascade * KING_PAIR_BONUS
    }

    return minOf(1.0, tossInValue / MAX_TOSS_IN_VALUE)
}

private const val DEFAULT_UNKNOWN_VALUE = 5
private const val CASCADE_MULTIPLIER = 3
private const val HIGH_VALUE_THRESHOLD = 10
private const val HIGH_VALUE_BONUS = 5
private const val KING_PAIR_BONUS = 10
private const val MAX_TOSS_IN_VALUE = 60.0

/**
 * Normalisation ranges for [evaluateRelativePosition].
 *
 * These map a raw advantage onto 0-1. The offsets say where "level" sits — being 10 points
 * behind the field still scores above zero, because the game is long — and the spans say how
 * big a lead has to be before it counts as won.
 */
private const val SCORE_ADVANTAGE_OFFSET = 10
private const val SCORE_ADVANTAGE_SPAN = 30
private const val CARD_ADVANTAGE_OFFSET = 2
private const val CARD_ADVANTAGE_SPAN = 5
private const val BEST_RIVAL_OFFSET = 5
private const val BEST_RIVAL_SPAN = 25

/** Where the bot stands against the table, on score and on hand size. */
fun evaluateRelativePosition(state: MctsGameState, botPlayer: MctsPlayerState): Double {
    val opponents = state.players.filter { it.id != botPlayer.id }
    if (opponents.isEmpty()) return 0.0

    val averageOpponentScore = opponents.sumOf { it.score } / opponents.size
    val bestOpponentScore = opponents.minOf { it.score }
    val averageOpponentCards = opponents.sumOf { it.cardCount }.toDouble() / opponents.size

    val scoreComponent = ((averageOpponentScore - botPlayer.score + SCORE_ADVANTAGE_OFFSET) /
        SCORE_ADVANTAGE_SPAN).coerceIn(0.0, 1.0)
    val cardComponent = ((averageOpponentCards - botPlayer.cardCount + CARD_ADVANTAGE_OFFSET) /
        CARD_ADVANTAGE_SPAN).coerceIn(0.0, 1.0)
    // Being ahead of the *field* is worth less than being ahead of whoever is actually winning.
    val competitive = ((bestOpponentScore - botPlayer.score + BEST_RIVAL_OFFSET) /
        BEST_RIVAL_SPAN).coerceIn(0.0, 1.0)

    return scoreComponent * POSITION_SCORE_WEIGHT +
        cardComponent * POSITION_CARD_WEIGHT +
        competitive * POSITION_RIVAL_WEIGHT
}

/** Score matters most, then how many cards are left to shed, then the closest rival. */
private const val POSITION_SCORE_WEIGHT = 0.5
private const val POSITION_CARD_WEIGHT = 0.3
private const val POSITION_RIVAL_WEIGHT = 0.2

/**
 * Action cards in hand, weighted by how much each action is actually worth.
 *
 * The `when` below is a tuning table and is left as literals on purpose: a name per rank
 * would say no more than the rank already does, and the table is easier to re-tune when the
 * numbers sit beside their cases.
 */
@Suppress("MagicNumber")
fun evaluateActionCardValue(botPlayer: MctsPlayerState): Double {
    var actionValue = 0.0
    val rankCounts = mutableMapOf<Rank, Int>()

    for (position in 0 until botPlayer.cardCount) {
        val memory = botPlayer.knownCards[position] ?: continue
        if (memory.confidence <= TRUSTED_CONFIDENCE) continue

        val rank = memory.card.rank
        rankCounts[rank] = (rankCounts[rank] ?: 0) + 1
        if (!rank.isActionable()) continue

        actionValue += when (rank) {
            Rank.KING -> 15.0
            Rank.QUEEN, Rank.JACK -> 10.0
            Rank.NINE, Rank.TEN -> 6.0
            Rank.SEVEN, Rank.EIGHT -> 4.0
            Rank.ACE -> 3.0
            else -> 0.0
        }
    }

    // A King can declare a rank the bot already holds a pair of, turning one action into a
    // cascade — so the two together are worth more than either apart.
    val pairCount = rankCounts.count { it.value >= 2 }
    if (rankCounts.containsKey(Rank.KING) && pairCount > 0) {
        actionValue += pairCount * KING_SYNERGY_BONUS
    }

    return minOf(1.0, actionValue / MAX_ACTION_VALUE)
}

private const val KING_SYNERGY_BONUS = 8
private const val MAX_ACTION_VALUE = 50.0

/** How much the bot knows, weighted towards its own hand — which is what it can act on. */
@Suppress("MagicNumber")
fun evaluateInformationAdvantage(state: MctsGameState, botPlayer: MctsPlayerState): Double {
    val ownKnowledge =
        if (botPlayer.cardCount > 0) botPlayer.knownCards.size.toDouble() / botPlayer.cardCount
        else 0.0

    val opponents = state.players.filter { it.id != botPlayer.id }
    val totalOpponentCards = opponents.sumOf { it.cardCount }
    val knownOpponentCards = opponents.sumOf { it.knownCards.size }
    val opponentKnowledge =
        if (totalOpponentCards > 0) knownOpponentCards.toDouble() / totalOpponentCards else 0.0

    return ownKnowledge * 0.6 + opponentKnowledge * 0.4
}

/** Inverted danger: an opponent who is shorter, lower-scoring or unreadable is a threat. */
@Suppress("MagicNumber")
fun evaluateThreatLevel(state: MctsGameState, botPlayer: MctsPlayerState): Double {
    var threat = 0.0

    for (opponent in state.players.filter { it.id != botPlayer.id }) {
        if (opponent.cardCount < botPlayer.cardCount) {
            threat += (botPlayer.cardCount - opponent.cardCount) * 0.1
        }
        if (opponent.score < botPlayer.score) {
            threat += (botPlayer.score - opponent.score) / 40
        }
        val unknownRatio =
            if (opponent.cardCount > 0) {
                (opponent.cardCount - opponent.knownCards.size).toDouble() / opponent.cardCount
            } else {
                0.0
            }
        threat += unknownRatio * 0.08
    }

    return maxOf(0.0, 1 - threat)
}
